lib/cylc/remote.py

```diff
@@ -283,3 +298,586 @@ def execute(self, dry_run=False, forward_x11=False, abort_if=None):
             sys.exit(r'ERROR: remote command failed %d' % res)
         else:
             return True
+
+
+class EmptyHostList(Exception):
+    """Exception to be raised if there are no hosts considered valid for
+       appointment given the 'run hosts' and 'run host select' global
+       configuration specifications."""
+
+    def __str__(self):
+        return "No hosts currently compatible with the global configuration."
+
+
+class HostAppointer(object):
+    """Determine the one host most suitable to (re-)run a suite on from all
+       'run hosts' and given the 'run host selection' ranking options
+       as specified (else taken as default) in the global configuration.
+    """
+
+    CMD_BASE = "get-host-metrics"  # 'cylc' prepended by remote_cylc_cmd.
+    try:
+        HOSTS = glbl_cfg().get(['suite servers', 'run hosts'])
+        RANK_METHOD = glbl_cfg().get(
+            ['suite servers', 'run host select', 'rank'])
+        THRESHOLDS = glbl_cfg().get(
+            ['suite servers', 'run host select', 'thresholds'])
+    except KeyError:  # Catch for unittest which updates these internally.
+        HOSTS = []
+        RANK_METHOD = 'random'
+        THRESHOLDS = None
+
+    def __init__(self):
+        self.use_disk_path = "/"
+        self.opt_mapping = {
+            "load:1": ("--load", ["load", "1 min"]),
+            "load:5": ("--load", ["load", "5 min"]),
+            "load:15": ("--load", ["load", "15 min"]),
+            "memory": ("--memory", ["memory"]),
+            "disk-space": ("--disk-space=" + self.use_disk_path,
+                           ["disk-space", self.use_disk_path])
+        }
+        self.PARSED_THRESHOLDS = self.parse_thresholds(self.THRESHOLDS)
+
+    @staticmethod
+    def parse_thresholds(raw_thresholds_spec):
+        if not raw_thresholds_spec:
+            return {}
+        valid_thresholds = {}
+        for threshold in raw_thresholds_spec.split(';'):
+            threshold = threshold.strip()
+            try:
+                measure, cutoff = threshold.split(' ')
+            except (AttributeError, ValueError):
+                print "Invalid threshold component '%s'." % threshold
+                raise
+            try:
+                if measure in ("load:1", "load:5", "load:15"):
+                    cutoff = float(cutoff)
+                elif measure in ("memory", "disk-space"):
+                    cutoff = int(cutoff)
+                else:
+                    raise AttributeError("Invalid threshold measure '%s'." %
+                                         measure)
+            except ValueError:
+                print "Threshold value '%s' of wrong type." % cutoff
+                raise
```

Comment:
Add error string to `ValueError` or print to STDERR.

Reply:
Sorry, my error handling for this PR is often not correct, as spotted. Thanks for pointing out these instances.